File: apple/HDHROpenKit/Sources/HDHROpenKit/Networking/APIClient+AI.swift
import Foundation

extension APIClient {
    // MARK: - AI Assistant APIs

    public func sendAIChat(
        request: AIChatRequest,
        onEvent: @Sendable @escaping (AIStreamEvent) -> Void
    ) async throws {
        guard let url = URL(string: APIEndpoints.aiChat(), relativeTo: baseURL) else {
            throw APIError.invalidURL
        }

        var urlRequest = URLRequest(url: url)
        urlRequest.httpMethod = "POST"
        urlRequest.setValue("application/json", forHTTPHeaderField: "Content-Type")

        if let token = bearerToken {
            urlRequest.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        if let devId = deviceId {
            urlRequest.setValue(devId, forHTTPHeaderField: "X-Device-Id")
        }

        urlRequest.httpBody = try jsonEncoder.encode(request)

        let (asyncBytes, response) = try await session.bytes(for: urlRequest)
        guard let httpResponse = response as? HTTPURLResponse else {
            throw APIError.networkError("Invalid server response")
        }

        if httpResponse.statusCode >= 400 {
            throw APIError.serverError(statusCode: httpResponse.statusCode, message: "AI chat failed with status code \(httpResponse.statusCode)")
        }

        for try await line in asyncBytes.lines {
            if line.hasPrefix("data: ") {
                let jsonString = String(line.dropFirst(6)).trimmingCharacters(in: .whitespacesAndNewlines)
                if jsonString.isEmpty { continue }
                if let data = jsonString.data(using: .utf8) {
                    do {
                        let event = try jsonDecoder.decode(AIStreamEvent.self, from: data)
                        onEvent(event)
                    } catch {
                        Log.network.error("Error decoding AI stream event: \(error.localizedDescription)")
                    }
                }
            }
        }
    }

    public func testAIConnection(payload: [String: AnyCodable] = [:]) async throws -> NetworkTestConnectionResult {
        try await request(path: APIEndpoints.aiTestConnection(), method: "POST", body: payload)
    }

    public func listAIModels(payload: [String: AnyCodable] = [:]) async throws -> AIListModelsResponse {
        try await request(path: APIEndpoints.aiListModels(), method: "POST", body: payload)
    }

    public func confirmAIAction(actionId: String) async throws -> AIConfirmActionResponse {
        try await request(path: APIEndpoints.aiConfirmAction(actionId), method: "POST")
    }

    public func cancelAIAction(actionId: String) async throws -> AICancelActionResponse {
        try await request(path: APIEndpoints.aiCancelAction(actionId), method: "POST")
    }
}

File: android/core/src/main/kotlin/org/hdhropen/kit/networking/APIClient+AI.kt
package org.hdhropen.kit.networking

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.JsonElement
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.hdhropen.kit.models.*
import org.hdhropen.kit.utilities.Log

private val aiJsonMediaType = "application/json; charset=utf-8".toMediaType()

suspend fun APIClient.sendAIChat(
    request: AIChatRequest,
    onEvent: (AIStreamEvent) -> Unit
) = withContext(Dispatchers.IO) {
    val fullUrl = if (baseURL.endsWith("/")) {
        baseURL + APIEndpoints.aiChat().removePrefix("/")
    } else {
        baseURL + APIEndpoints.aiChat()
    }

    val jsonString = json.encodeToString(request)
    val body = jsonString.toRequestBody(aiJsonMediaType)

    val reqBuilder = Request.Builder()
        .url(fullUrl)
        .post(body)

    bearerToken?.let { reqBuilder.header("Authorization", "Bearer $it") }
    deviceId?.let { reqBuilder.header("X-Device-Id", it) }

    val response = try {
        httpClient.newCall(reqBuilder.build()).execute()
    } catch (e: Exception) {
        throw APIError.NetworkError(e.localizedMessage ?: "Failed to connect to AI assistant")
    }

    if (!response.isSuccessful) {
        val code = response.code
        val errorMsg = response.body?.string() ?: "HTTP $code"
        throw APIError.ServerError(code, errorMsg)
    }

    val stream = response.body?.byteStream() ?: return@withContext
    stream.bufferedReader().useLines { lines ->
        for (line in lines) {
            if (line.startsWith("data: ")) {
                val dataContent = line.removePrefix("data: ").trim()
                if (dataContent.isNotEmpty()) {
                    try {
                        val event = json.decodeFromString<AIStreamEvent>(dataContent)
                        onEvent(event)
                    } catch (e: Exception) {
                        Log.network.error("Error decoding AI stream event: ${e.message}")
                    }
                }
            }
        }
    }
}

suspend fun APIClient.testAIConnection(payload: Map<String, JsonElement> = emptyMap()): NetworkTestConnectionResult {
    return request(path = APIEndpoints.aiTestConnection(), method = "POST", body = payload)
}

suspend fun APIClient.listAIModels(payload: Map<String, JsonElement> = emptyMap()): AIListModelsResponse {
    return request(path = APIEndpoints.aiListModels(), method = "POST", body = payload)
}

suspend fun APIClient.confirmAIAction(actionId: String): AIConfirmActionResponse {
    return request(path = APIEndpoints.aiConfirmAction(actionId), method = "POST")
}

suspend fun APIClient.cancelAIAction(actionId: String): AICancelActionResponse {
    return request(path = APIEndpoints.aiCancelAction(actionId), method = "POST")
}
